<div style="text-align:left;font-size:2em"><span style="font-weight:bolder;font-size:1.25em">SP2273 | Learning Portfolio</span><br><br><span style="font-weight:bold;color:darkred">Using Jupyter (Good), Exercises</span></div>

## Exercise 1

In [1]:
# while True:
#     print('*', end='')

# did interrupt kernel (KeyboardInterrupt error).
# While loop commented due to an issue (long loading time) while opening the file.

## Exercise 2

In [7]:
print = 'PRINT'
print('Hello World')

TypeError: 'str' object is not callable

## Exercise 3

In [3]:
%%timeit
squares = [i**2 for i in range(10)]

# This code will be faster as it is only one line of code and can be read faster than three lines of code.

590 ns ± 5.45 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


In [4]:
%%timeit
squares = list()
for i in range(10):
    squares.append(i**2)

639 ns ± 2.76 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


**Explaining the difference between the timings**

590 and 639 represent the mean time of running the function, i.e, over 1,000,000 iterations, the code took 590ns and 639ns to run.

5.45 and 2.76 represent the average deviation of run time from the mean, i.e, over 1,000,000 iterations, the code had a variation of run time by 5.45ns and 2.76ns. The fact that the second code had a shorter variation means it was more consistent in its execution in comparison to code 1.

Code 1 is faster due to (a) comprehension of the list, i.e, it is a single operation (while the 2nd code has multiple operations) and (b) there is no loop in code 1 (the for loop may have increased the run time in code 2)

In [5]:
squares = [i**2 for i in range(10)]
print(squares)

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]


In [6]:
squares = list()
for i in range(10):
    squares.append(i**2)

## Footnotes